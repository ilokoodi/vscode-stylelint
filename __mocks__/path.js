'use strict';

const path = jest.requireActual('path');

/**
 * Mock platform.
 * @type {'posix' | 'win32'}
 */
let mockPlatform = 'posix';

/**
 * Sets the mock platform.
 * @param {'posix' | 'win32'} platform
 * @returns {void}
 */
const __mockPlatform = (platform) => {
	mockPlatform = platform;
};

const pathProxy = new Proxy(path, {
	get(_, name) {
		if (name === '__mockPlatform') {
			return __mockPlatform;
		}

		if (mockPlatform === 'win32') {
			if (typeof path.win32[name] === 'function') {
				return jest.fn(path.win32[name]);
			}

			return path.win32[name];
		}

		if (typeof path.posix[name] === 'function') {
			return jest.fn(path.posix[name]);
		}

		return path.posix[name];
	},
});

module.exports = pathProxy;
