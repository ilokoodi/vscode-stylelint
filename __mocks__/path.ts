const path = jest.requireActual('path');

/**
 * Mock platform.
 */
let mockPlatform: 'posix' | 'win32' | undefined = undefined;

/**
 * Sets the mock platform. A value of `undefined` will use the actual platform.
 */
const __mockPlatform = (platform?: 'posix' | 'win32'): void => {
	mockPlatform = platform;
};

const mockedFns: {
	default: Record<string | symbol, () => unknown>;
	posix: Record<string | symbol, () => unknown>;
	win32: Record<string | symbol, () => unknown>;
} = {
	default: {},
	posix: {},
	win32: {},
};

const pathProxy = new Proxy(path, {
	get(_, name) {
		if (name === '__mockPlatform') {
			return __mockPlatform;
		}

		if (!mockPlatform) {
			if (typeof path[name] === 'function') {
				if (mockedFns.default[name]) {
					return mockedFns.default[name];
				}

				mockedFns.default[name] = jest.fn(path[name]);

				return mockedFns.default[name];
			}

			return path[name];
		}

		if (mockPlatform === 'win32') {
			if (typeof path.win32[name] === 'function') {
				if (mockedFns.win32[name]) {
					return mockedFns.win32[name];
				}

				mockedFns.win32[name] = jest.fn(path.win32[name]);

				return mockedFns.win32[name];
			}

			return path.win32[name];
		}

		if (typeof path.posix[name] === 'function') {
			if (mockedFns.posix[name]) {
				return mockedFns.posix[name];
			}

			mockedFns.posix[name] = jest.fn(path.posix[name]);

			return mockedFns.posix[name];
		}

		return path.posix[name];
	},
});

export = pathProxy;
