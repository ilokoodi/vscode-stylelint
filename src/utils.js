const fs = require('fs').promises;
const path = require('path');

/**
 * Walks up the file tree from the current directory until it finds a
 * directory containing `package.json`. Resolves to `undefined` if no such
 * directory is found.
 * @param {string} directory The directory to start from.
 * @returns {Promise<string | undefined>}
 */
async function resolvePackageDirectory(directory) {
	let currentDirectory = directory;

	// eslint-disable-next-line no-constant-condition
	while (true) {
		const manifestPath = path.join(currentDirectory, 'package.json');

		try {
			await fs.access(manifestPath);

			return currentDirectory;
		} catch (error) {
			if (/** @type {{code?: string}} */ (error).code === 'ENOENT') {
				const parent = path.dirname(currentDirectory);

				if (!path.relative(parent, currentDirectory)) {
					return undefined;
				}

				currentDirectory = parent;
			} else {
				throw error;
			}
		}
	}
}

module.exports = {
	resolvePackageDirectory,
};
