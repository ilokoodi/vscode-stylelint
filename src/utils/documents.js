'use strict';

const pathIsInside = require('path-is-inside');
const { URI } = require('vscode-uri');

/**
 * @param {lsp.TextDocument} document
 * @param {lsp.Connection} [connection]
 * @returns {Promise<string | undefined>}
 */
async function getWorkspaceFolder(document, connection) {
	const documentPath = URI.parse(document.uri).fsPath;

	if (documentPath) {
		const workspaceFolders = await connection?.workspace.getWorkspaceFolders();

		if (workspaceFolders) {
			for (const { uri } of workspaceFolders) {
				const workspacePath = URI.parse(uri).fsPath;

				if (pathIsInside(documentPath, workspacePath)) {
					return workspacePath;
				}
			}
		}
	}

	return undefined;
}

module.exports = {
	getWorkspaceFolder,
};
