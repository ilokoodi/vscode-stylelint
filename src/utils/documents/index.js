'use strict';

module.exports = {
	...require('./create-text-edits'),
	...require('./get-disable-type'),
	...require('./get-fixes'),
	...require('./get-workspace-folder'),
};
