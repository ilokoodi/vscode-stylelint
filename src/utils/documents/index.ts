export * from './create-text-edits';
export * from './get-disable-type';
export * from './get-fixes';
export * from './get-workspace-folder';
