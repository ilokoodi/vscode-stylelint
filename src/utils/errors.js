'use strict';

const { serializeError } = require('serialize-error');
const { isIterable } = require('./iterables');

/**
 * Takes an object with errors, returning a new object in which each error has
 * been replaced by a serializable object with the error's properties. Iterables
 * that aren't arrays, maps, or sets are converted to arrays.
 * @template T
 * @template {{[K in keyof T]: T[K]}} R
 * @param {T} object The object with errors.
 * @returns {R} The object with each error replaced by a serializable object.
 */
function serializeErrors(object) {
	/**
	 * @template TInner
	 * @template {{[KInner in keyof TInner]: TInner[KInner]}} RInner
	 * @param {TInner} obj The object with errors.
	 * @param {WeakMap<any, any>} visited The set of objects that have already been visited.
	 * @returns {RInner} The object with each error replaced by a serializable object.
	 */
	const serializeInner = (obj, visited) => {
		if (!obj || typeof obj !== 'object') {
			return /** @type {RInner} */ (obj);
		}

		if (visited.has(obj)) {
			return visited.get(obj);
		}

		if (obj instanceof Error) {
			const result = /** @type {RInner} */ (/** @type {unknown} */ (serializeError(obj)));

			visited.set(obj, result);

			return result;
		}

		if (obj instanceof Map) {
			const result = new Map();

			visited.set(obj, result);

			for (const [key, value] of /** @type {Map<any, any>} */ (/** @type {unknown} */ (obj))) {
				const serializedKey = serializeInner(key, visited);
				const serializedValue = serializeInner(value, visited);

				if (key && typeof key === 'object') {
					visited.set(key, serializedKey);
				}

				if (value && typeof value === 'object') {
					visited.set(value, serializedValue);
				}

				result.set(serializedKey, serializedValue);
			}

			return /** @type {RInner} */ (/** @type {unknown} */ (result));
		}

		if (obj instanceof Set) {
			const result = new Set();

			visited.set(obj, result);

			for (const value of /** @type {Set<any>} */ (/** @type {unknown} */ (obj))) {
				if (!value || typeof value !== 'object') {
					result.add(value);
					continue;
				}

				const serializedValue = serializeInner(value, visited);

				visited.set(value, serializedValue);

				result.add(serializedValue);
			}

			return /** @type {RInner} */ (/** @type {unknown} */ (result));
		}

		if (isIterable(obj)) {
			/** @type {any[]} */
			const result = [];

			visited.set(obj, result);

			for (const value of /** @type {Iterable<any>} */ (/** @type {unknown} */ (obj))) {
				result.push(serializeInner(value, visited));
			}

			return /** @type {RInner} */ (/** @type {unknown} */ (result));
		}

		visited.set(obj, '[Circular]');

		const serializedObj = /** @type {RInner} */ (
			Object.fromEntries(
				Object.entries(obj).map(([key, value]) => {
					if (!value || typeof value !== 'object') {
						return [key, value];
					}

					if (visited.has(value)) {
						return [key, visited.get(value)];
					}

					if (value instanceof Error) {
						const serialized = serializeError(value);

						visited.set(value, serialized);

						return [key, serialized];
					}

					const result = serializeInner(value, visited);

					visited.set(value, result);

					return [key, result];
				}),
			)
		);

		visited.set(obj, serializedObj);

		return serializedObj;
	};

	return serializeInner(object, new WeakMap());
}

module.exports = {
	serializeErrors,
};
