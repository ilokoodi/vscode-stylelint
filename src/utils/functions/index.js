'use strict';

module.exports = {
	...require('./get-first-return-value'),
	...require('./lazy-call'),
};
