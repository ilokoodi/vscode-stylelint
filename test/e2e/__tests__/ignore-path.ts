import path from 'path';
import { languages, Diagnostic, TextDocument } from 'vscode';
import { normalizeDiagnostic } from '../utils';

describe('"stylelint.ignorePath" setting', () => {
	it('should be respected for files that are ignored', async () => {
		const { document } = await openDocument(path.resolve(workspaceDir, 'ignore-path/ignored.css'));

		// If a file is ignored, diagnostics are returned as an empty object ([]), which never
		// resolves with the current implementation of `waitForDiagnostics`. This promise uses
		// a listener to resolve with the empty diagnostics when they are available.
		const waitForEmptyDiagnostics = (document: TextDocument) =>
			new Promise<Diagnostic[]>((resolve) => {
				const onDidChangeDiagnostics = languages.onDidChangeDiagnostics(
					(diagnosticsChangeEvent) => {
						if (
							diagnosticsChangeEvent.uris.filter((uri) => uri.path === document.uri.path).length > 0
						) {
							resolve(languages.getDiagnostics(document.uri));
							onDidChangeDiagnostics.dispose();
						}
					},
				);
			});

		const diagnostics = await waitForEmptyDiagnostics(document);
		expect(diagnostics).toEqual([]);
	});

	it('should lint files that are not ignored', async () => {
		const { document } = await openDocument(
			path.resolve(workspaceDir, 'ignore-path/not-ignored.css'),
		);
		const diagnostics = await waitForDiagnostics(document);

		expect(diagnostics.map(normalizeDiagnostic)).toMatchSnapshot();
	});
});
