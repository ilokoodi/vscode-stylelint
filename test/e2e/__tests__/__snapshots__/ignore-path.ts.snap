// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"stylelint.ignorePath" setting should lint files that are not ignored 1`] = `
Array [
  Object {
    "code": "indentation",
    "codeDescription": Object {
      "href": "https://stylelint.io/user-guide/rules/list/indentation",
    },
    "message": "Expected indentation of 4 spaces (indentation)",
    "range": Object {
      "end": Object {
        "character": 3,
        "line": 2,
      },
      "start": Object {
        "character": 2,
        "line": 2,
      },
    },
    "severity": 1,
    "source": "Stylelint",
  },
]
`;
